src/Functions/printf.cpp

```diff
@@ -0,0 +1,309 @@
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Functions/formatString.h>
+#include <IO/WriteHelpers.h>
+
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <fmt/format.h>
+#include <fmt/printf.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int ILLEGAL_COLUMN;
+extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+namespace
+{
+
+class FunctionPrintf : public IFunction
+{
+private:
+    ContextPtr context;
+    FunctionOverloadResolverPtr function_concat;
+
+    struct Instruction
+    {
+        std::string_view format;
+        size_t rows;
+        bool is_literal; /// format is literal string without any argument
+        ColumnWithTypeAndName input; /// Only used when is_literal is false
+
+        ColumnWithTypeAndName execute()
+        {
+            if (is_literal)
+                return executeLiteral(format);
+            else if (isColumnConst(*input.column))
+                return executeConstant(input);
+            else
+                return executeNonconstant(input);
+        }
+
+        String toString() const
+        {
+            std::ostringstream oss; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+            oss << "format:" << format << ", rows:" << rows << ", is_literal:" << is_literal << ", input:" << input.dumpStructure()
+                << std::endl;
+            return oss.str();
+        }
+
+    private:
+        ColumnWithTypeAndName executeLiteral(std::string_view literal)
+        {
+            ColumnWithTypeAndName res;
+            auto str_col = ColumnString::create();
+            str_col->insert(fmt::sprintf(literal));
+            res.column = ColumnConst::create(std::move(str_col), rows);
+            res.type = std::make_shared<DataTypeString>();
+            return res;
+        }
+
+        ColumnWithTypeAndName executeConstant(const ColumnWithTypeAndName & arg)
+        {
+            ColumnWithTypeAndName tmp_arg = arg;
+            const auto & const_col = static_cast<const ColumnConst &>(*arg.column);
+            tmp_arg.column = const_col.getDataColumnPtr();
+
+            ColumnWithTypeAndName tmp_res = executeNonconstant(tmp_arg);
+            return ColumnWithTypeAndName{ColumnConst::create(tmp_res.column, arg.column->size()), tmp_res.type, tmp_res.name};
+        }
+
+        ColumnWithTypeAndName executeNonconstant(const ColumnWithTypeAndName & arg)
+        {
+            size_t size = arg.column->size();
+            auto res_col = ColumnString::create();
+            auto & res_str = static_cast<ColumnString &>(*res_col);
+            auto & res_offsets = res_str.getOffsets();
+            auto & res_chars = res_str.getChars();
+            res_offsets.reserve_exact(size);
+            res_chars.reserve(format.size() * size * 2);
+
+            String s;
+            WhichDataType which(arg.type);
+
+#define EXECUTE_BY_TYPE(IS_TYPE, GET_TYPE) \
```

Comment:
Can we rewrite it with template function? (maybe casting to ColumnVector Or ColumnString). I'm not strictly against implementing it with macros but if another solution is possible, I'd prefer it

Reply:
done.